libraries/adaptive-expressions/src/functionUtils.ts

```diff
@@ -768,14 +770,43 @@ export class FunctionUtils {
             return obj1 == null && obj2 == null;
         }
 
-        if (Array.isArray(obj1) && obj1.length === 0 && Array.isArray(obj2) && obj2.length === 0) {
-            return true;
+        // Array Comparison
+        if (Array.isArray(obj1) && Array.isArray(obj2)) {
+            if (obj1.length !== obj2.length) {
+                return false;
+            }
+
+            let isEqual = true;
+            for (let i = 0; i < obj1.length; i++) {
+                if (!FunctionUtils.commonEquals(obj1[i], obj2[i])) {
+                    isEqual = false;
+                    break;
+                }
+            }
+            return isEqual;
         }
 
-        if (FunctionUtils.getPropertyCount(obj1) === 0 && FunctionUtils.getPropertyCount(obj2) === 0) {
-            return true;
+        // Object Comparison
+        const propertyCountOfObj1 = FunctionUtils.getPropertyCount(obj1);
+        const propertyCountOfObj2 = FunctionUtils.getPropertyCount(obj2);
+        if (propertyCountOfObj1 >= 0 && propertyCountOfObj2 >= 0) {
+            if (propertyCountOfObj1 !== propertyCountOfObj2) {
+                return false;
+            }
+            let jsonObj1 = obj1;
+            let jsonObj2 = obj2;
+            if (obj1 instanceof Map) {
```

Comment:
Maybe this should be pushed into the `convertToObj` method, so these lines become `const jsonObj1 = FunctionUtils.convertToObj(obj1);`